File: frontend/app_mentorly/lib/features/auth/controllers/authController.dart
import '../../../core/services/authService.dart';
import '../models/userModel.dart';

// Controla o estado de login/cadastro, chamado pelas telas de auth
class AuthController {
  final AuthService _authService = AuthService();

  UserModel? usuarioLogado;
  String? erro;

  Future<bool> login(String email, String senha) async {
    try {
      usuarioLogado = await _authService.loginCoordenacao(email, senha);
      return true;
    } catch (e) {
      erro = e.toString();
      return false;
    }
  }

  Future<bool> loginProfessor({
    required String email,
    required String senha,
  }) async {
    try {
      usuarioLogado = await _authService.loginProfessor(email: email, senha: senha);
      return true;
    } catch (e) {
      erro = e.toString();
      return false;
    }
  }

  Future<bool> cadastrar({
    required String nome,
    required String email,
    required String senha,
    required String telefone,
  }) async {
    try {
      usuarioLogado = await _authService.cadastrarCoordenacao(
        nome: nome,
        email: email,
        senha: senha,
        telefone: telefone,
      );
      return true;
    } catch (e) {
      erro = e.toString();
      return false;
    }
  }

  Future<bool> criarSenhaProfessor({
    required String email,
    required String senha,
    required String confirmarSenha,
    required String token,
  }) async {
    try {
      usuarioLogado = await _authService.criarSenhaProfessor(
        email: email,
        senha: senha,
        confirmarSenha: confirmarSenha,
        token: token,
      );
      return true;
    } catch (e) {
      erro = e.toString();
      return false;
    }
  }

  Future<bool> confirmarCodigo(String email, String codigo) {
    return _authService.confirmarCodigoDuasEtapas(email, codigo);
  }
}

File: backend/controllers/auth_controller.py
"""Controller de autenticacao: so HTTP, nada de SQL nem regra de negocio.

O contrato de resposta ({token, usuario}) e o que o AuthService do Flutter
(core/services/authService.dart) ja espera.
"""

from flask import jsonify, request

from services.auth.cadastro_coordenacao import CadastroCoordenacaoService
from services.auth.criar_senha_professor import CriarSenhaProfessorService
from services.auth.login_coordenacao import LoginCoordenacaoService
from services.auth.login_professor import LoginProfessorService
from services.auth.verificacao_duas_etapas import (
    ConfirmarCodigoService,
    EnviarCodigoService,
)


class AuthController:
    def cadastro_coordenacao(self):
        dados = request.get_json(silent=True) or {}
        try:
            resultado = CadastroCoordenacaoService().execute(
                dados.get("nome"),
                dados.get("email"),
                dados.get("senha"),
                dados.get("telefone"),
            )
        except ValueError as erro:
            return jsonify({"error": str(erro)}), 400
        return jsonify(resultado), 201

    def login_coordenacao(self):
        dados = request.get_json(silent=True) or {}
        try:
            resultado = LoginCoordenacaoService().execute(
                dados.get("email"), dados.get("senha")
            )
        except ValueError as erro:
            return jsonify({"error": str(erro)}), 400
        except PermissionError as erro:
            return jsonify({"error": str(erro)}), 401
        return jsonify(resultado)

    def login_professor(self):
        dados = request.get_json(silent=True) or {}
        try:
            resultado = LoginProfessorService().execute(
                dados.get("email"), dados.get("senha")
            )
        except ValueError as erro:
            return jsonify({"error": str(erro)}), 400
        except PermissionError as erro:
            return jsonify({"error": str(erro)}), 401
        return jsonify(resultado)

    def criar_senha_professor(self):
        dados = request.get_json(silent=True) or {}
        try:
            resultado = CriarSenhaProfessorService().execute(
                dados.get("email"), dados.get("senha"), dados.get("token")
            )
        except ValueError as erro:
            return jsonify({"error": str(erro)}), 400
        except PermissionError as erro:
            return jsonify({"error": str(erro)}), 403
        return jsonify(resultado)

    def enviar_codigo(self):
        dados = request.get_json(silent=True) or {}
        try:
            resultado = EnviarCodigoService().execute(dados.get("email"))
        except ValueError as erro:
            return jsonify({"error": str(erro)}), 400
        return jsonify(resultado)

    def confirmar_codigo(self):
        dados = request.get_json(silent=True) or {}
        try:
            resultado = ConfirmarCodigoService().execute(
                dados.get("email"), dados.get("codigo")
            )
        except ValueError as erro:
            return jsonify({"error": str(erro)}), 400
        return jsonify(resultado)
